### PR TITLE
connection: wake 'stopped' streams on stream finish events

### PR DESCRIPTION
**Pull request by bmwill**
In the event that the SendStream Stopped future is polled once and then dropped before the remote side has stopped the stream, and instead the stream has been finished, the waker registered in the connection stopped map is never removed and is effectively "leaked" until the connection is closed. This can lead to a large amount of memory being retained when a connection is very long lived and many streams are used over the life of that connection.

This particular issue can be seen with this type of usage of SendStream::stopped: https://github.com/bmwill/anemo/blob/main/crates/anemo/src/network/request_handler.rs#L170-L177.

This is one such way to fix this memory retention, although it may not be the project's preferred way. Another option would potentially be to remove the waker from the hash map in a `Drop` impl for the `Stopped` type.

### CODE REVIEW

**Comment by bmwill** [comment]
> Another option would potentially be to remove the waker from the hash map in a Drop impl for the Stopped type.

Given `Stopped` retains an `&mut SendStream` internally, i'm thinking that using a Drop impl may actually be a preferable, more foolproof solution.

File: quinn/src/send_stream.rs [hunk 1]
```diff
@@ -307,6 +307,19 @@ impl Future for Stopped<'_> {
...
+impl<'a> Drop for Stopped<'a> {
+    fn drop(&mut self) {
+        let mut conn = self.stream.conn.state.lock("Stopped::drop");
+
+        if let Some(stopped) = conn.stopped.remove(&self.stream.stream) {
+            // Given this is the future that would have registered the waker we may not need to
+            // actually call Waker::wake, calling here may result in a spurious wake up but at
+            // least we can be certain that there isn't a task that will never be woken up.
+            stopped.wake();
+        }
```
Comment (Ralith) [inline quinn/src/send_stream.rs:319]: Is there a specific case of a never-woken task that you envision? I'm having trouble coming up with any cases where calling wake here is necessary.

**Review by Ralith — COMMENTED** [review]
Great catch, thanks!

File: quinn/src/send_stream.rs (hunk 1 above)
Comment (bmwill) [reply quinn/src/send_stream.rs:319]: I don't have a specific case, and would be completely comfortable just ignoring the removed waker. Just let me know what you'd prefer we do here and i'll make the change.

Comment (Ralith) [reply quinn/src/send_stream.rs:319]: In the interest of maintainability, let's ignore it then.

**Review by Ralith — APPROVED** [review]
Thanks!

**Pull request merged**